### PR TITLE
[codegen][Go] Fix compilation error of generated go code when schema is free form object

### PR DESCRIPTION
**Pull request by sebastien-rosset**
**codegen:**
1. Add helper function isAnyTypeSchema()
1. Add code comments to clearly differentiate between free-form object (which must be a map of string to value) and any type value (which can be array, object, string, integer...).
1. Remove unused min/max fields in CodegenProperty
1. Add code comments in CodegenProperty
1. Fix equals() method in CodegenProperty. The deprecated attribute was a comparison to itself

**golang:**
When a schema is Free Form object, the go-experimental generator sets the golang dataType to **NullableMap[string]interface{}**, which is not a valid go type. This causes a compilation failure of the generated code.

This PR is a fix for this problem.

```yaml
components:
  schema:
        stuff:
            type: object
            nullable: true
```

Issue: #5387

From this PR, it's clear there is a different interpretation of the OAS spec between the go-experimental generator versus the python-experimental generator. This is a problem that needs to be addressed to get consistency across languages.

* In the generated go client, the only accepted type is a map of string to values
* In the generated python client, any valid JSON document will be accepted.

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### CODE REVIEW

**Comment by sebastien-rosset** [comment]
@spacether

**Comment by spacether** [comment]
Does interface allow array and map types?

**Comment by sebastien-rosset** [comment]
> Does interface allow array and map types?

you mean the golang **interface{}** type? If so yes, but see referenced issue #5387 for details. The way I understand the spec, a "type: object" is only an an unordered set of properties mapping a string to an instance.

File: samples/openapi3/client/petstore/go-experimental/go-petstore/api/openapi.yaml [hunk 1]
```diff
@@ -1303,6 +1303,7 @@ components:
         id: 0
         email: email
         username: username
+        stuff: '{}'
```
Comment (jirikuncar) [inline samples/openapi3/client/petstore/go-experimental/go-petstore/api/openapi.yaml]: IMHO it should be a JSON object not string '{}'.
```suggestion
        stuff: {}
```

Comment (sebastien-rosset) [reply samples/openapi3/client/petstore/go-experimental/go-petstore/api/openapi.yaml]: You are correct, but this is generated code. There must be another bug in the existing code that causes this issue. Let me check.

Comment (sebastien-rosset) [reply samples/openapi3/client/petstore/go-experimental/go-petstore/api/openapi.yaml]: This is generated from https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator/src/main/resources/go-experimental/openapi.mustache. There must be a problem with the content of the **openapi-yaml** tag.

Comment (sebastien-rosset) [reply samples/openapi3/client/petstore/go-experimental/go-petstore/api/openapi.yaml]: Never mind, the generated code is correct and must be within quotes. This is because this line is part of the generated example value.
   
https://github.com/OpenAPITools/openapi-generator/blob/40470a619d74022e850a711e7e4643ab1fced06a/samples/openapi3/client/petstore/go-experimental/go-petstore/api/openapi.yaml#L1297

Comment (sebastien-rosset) [reply samples/openapi3/client/petstore/go-experimental/go-petstore/api/openapi.yaml]: I am going to mark this issue resolved. Reopen if you have more questions.

**Comment by sebastien-rosset** [comment]
> Does interface allow array and map types?
> Those should be allowed per https://swagger.io/docs/specification/data-models/data-types/#any

Good point, thanks for catching that.

Case 1: the following schema must validate to a map of string to values (no other type allowed):
```yaml
components:
  schema:
        stuff:
            type: object
```

Case 2: the following schema must validate to the null value or a map of string to values:
```yaml
components:
  schema:
        stuff:
            type: object
            nullable: true
```

Case 3: the following schema can validate to any valid JSON document:
```yaml
components:
  schema:
        stuff: {}
```

File: samples/openapi3/client/petstore/go-experimental/go-petstore/docs/NullableClass.md [hunk 2]
```diff
@@ -10,12 +10,12 @@ Name | Type | Description | Notes
 **StringProp** | Pointer to **NullableString** |  | [optional] 
 **DateProp** | Pointer to **NullableString** |  | [optional] 
 **DatetimeProp** | Pointer to [**NullableTime**](time.Time.md) |  | [optional] 
-**ArrayNullableProp** | Pointer to [**[]map[string]interface{}**](map[string]interface{}.md) |  | [optional] 
-**ArrayAndItemsNullableProp** | Pointer to [**[]map[string]interface{}**](map[string]interface{}.md) |  | [optional] 
-**ArrayItemsNullable** | Pointer to [**[]map[string]interface{}**](map[string]interface{}.md) |  | [optional] 
-**ObjectNullableProp** | Pointer to [**map[string]map[string]interface{}**](map[string]interface{}.md) |  | [optional] 
-**ObjectAndItemsNullableProp** | Pointer to [**map[string]map[string]interface{}**](map[string]interface{}.md) |  | [optional] 
-**ObjectItemsNullable** | Pointer to [**map[string]map[string]interface{}**](map[string]interface{}.md) |  | [optional] 
+**ArrayNullableProp** | Pointer to **[]map[string]interface{}** |  | [optional]
```
Comment (sebastien-rosset) [inline samples/openapi3/client/petstore/go-experimental/go-petstore/docs/NullableClass.md]: This was a bug before this PR. There is no markdown file "map[string]interface{}.md"

File: samples/openapi3/client/petstore/python-experimental/petstore_api/models/user.py [hunk 3]
```diff
@@ -136,6 +144,10 @@ def __init__(self, _check_type=True, _from_server=False, _path_to_item=(), _conf
             password (str): [optional]  # noqa: E501
             phone (str): [optional]  # noqa: E501
             user_status (int): User Status. [optional]  # noqa: E501
+            arbitrary_object (bool, date, datetime, dict, float, int, list, str): test code generation for objects Value must be a map of strings to values. It cannot be the &#39;null&#39; value.. [optional]  # noqa: E501
```
Comment (sebastien-rosset) [inline samples/openapi3/client/petstore/python-experimental/petstore_api/models/user.py]: @spacether , these allowed types would have to be fixed in a separate PR.

Comment (spacether) [reply samples/openapi3/client/petstore/python-experimental/petstore_api/models/user.py]: Yup, this weekend I am busy but I can get to it in the future

**Comment by sebastien-rosset** [comment]
@antihax  @bvwells  @grokify @kemokemo @bkabrda

**Comment by bkabrda** [comment]
Hey 👋 could you please check out https://github.com/OpenAPITools/openapi-generator/pull/5414? I think one of us would need to modify their PR so that they would both work together correctly (also I'd like to hear your thoughts on that PR of mine, since you seem to be a user of go-experimental clients and this does some major changes to fix support for explicit nulls). Thanks!

**Comment by sebastien-rosset** [comment]
> Hey 👋 could you please check out #5414? I think one of us would need to modify their PR so that they would both work together correctly (also I'd like to hear your thoughts on that PR of mine, since you seem to be a user of go-experimental clients and this does some major changes to fix support for explicit nulls). Thanks!

I looked at your PR, and they don't seem to overlap much. Your PR is about nullable objects, my PR is about free-form objects and any type objects. Do you have something specific in mind?

**Comment by spacether** [comment]
@sebastien-rosset are there active reviewers in the go technical committee?

**Comment by sebastien-rosset** [comment]
> @sebastien-rosset are there active reviewers in the go technical committee?

Not as active as other languages like Python

**Comment by spacether** [comment]
> > @sebastien-rosset are there active reviewers in the go technical committee?
> 
> Not as active as other languages like Python

As far as I can tell I am the only active reviewer on the python technical committee at this time.
If you need help moving this forward please let me know. This coming week I should have some time.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractGoCodegen.java [hunk 4]
```diff
@@ -350,6 +375,9 @@ public String getSchemaType(Schema p) {
 
         if (ref != null && !ref.isEmpty()) {
             type = openAPIType;
+        } else if (openAPIType == "object" && ModelUtils.isAnyTypeSchema(p)) {
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractGoCodegen.java]: Should this be `"object".equals(openAPIType)` ?

Comment (sebastien-rosset) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractGoCodegen.java]: yes, thank you. Resolved

**Comment by bkabrda** [comment]
@sebastien-rosset you're right, I took a second look and it doesn't seem there's an overlap between our PRs. I was initially mistaken just by seeing something "nullable" in the code of your PR, but on a closer inspection, there's nothing problematic. Thanks for taking a look!

**Comment by sebastien-rosset** [comment]
@bkabrda, if you could review this PR, that would be great.

**Comment by wing328** [comment]
@sebastien-rosset can you please resolve the merge conflicts when you've time?

**Comment by sebastien-rosset** [comment]
> @sebastien-rosset can you please resolve the merge conflicts when you've time?

Done.

**Comment by sebastien-rosset** [comment]
After merge of #5414 , the generated code has compilation errors. TODO, fix.
```
go-petstore/model_user.go:344:42: o.ArbitraryNullableObject.Get undefined (type map[string]interface {} has no field or method Get)
go-petstore/model_user.go:348:35: o.ArbitraryNullableObject.Get undefined (type map[string]interface {} has no field or method Get)
go-petstore/model_user.go:358:34: o.ArbitraryNullableObject.Get undefined (type map[string]interface {} has no field or method Get)
go-petstore/model_user.go:358:67: o.ArbitraryNullableObject.IsSet undefined (type map[string]interface {} has no field or method IsSet)
go-petstore/model_user.go:363:42: o.ArbitraryNullableObject.IsSet undefined (type map[string]interface {} has no field or method IsSet)
go-petstore/model_user.go:372:27: o.ArbitraryNullableObject.Set undefined (type map[string]interface {} has no field or method Set)
go-petstore/model_user.go:376:27: o.ArbitraryNullableObject.Set undefined (type map[string]interface {} has no field or method Set)
go-petstore/model_user.go:381:27: o.ArbitraryNullableObject.Unset undefined (type map[string]interface {} has no field or method Unset)
go-petstore/model_user.go:418:45: o.ArbitraryNullableTypeValue.Get undefined (type interface {} is interface with no methods)
go-petstore/model_user.go:422:38: o.ArbitraryNullableTypeValue.Get undefined (type interface {} is interface with no methods)
go-petstore/model_user.go:422:38: too many errors
```

**Comment by bkabrda** [comment]
@sebastien-rosset sorry for that :/ I might have some spare time by the end of the week, so might be able to help fix your PR if you need.

**Comment by sebastien-rosset** [comment]
> @sebastien-rosset sorry for that :/ I might have some spare time by the end of the week, so might be able to help fix your PR if you need.

I have addressed the issues (I think).
